### PR TITLE
bug: move canBeConnected to PeerManager and check for potential overflow

### PR DESCRIPTION
**Pull request by vpavlin**
This PR moves `canBeConnected` to PeerManager. It also adds a check for input parameters for backoff calculation to make sure we do not hit overflow once the node is up and running - it checks that the max backoff is lower than 1 week and raises an exception otherwise.

It also adds tests to verify the potential overflow is caught at PeerManager initialization.

closes: #1665

### CODE REVIEW

**Comment by vpavlin** [comment]
@alrevuelta I could also see another approach where we add `maxFailedAttempts` to peerStore object, so that we do not need to pass it in each `canBeConnected` call. Let me know if I should change the PR

**Review by alrevuelta — CHANGES_REQUESTED** [review]
thanks for the PR!

Indeed, perhaps its better to add `maxFailedAttempts` to the peerstore, and avoid passing it in `canBeConnected`. Actually it may be also a good moment to do the same with `backoffFactor` and `initialBackoffInSec`. Would be very clean to have just `.canBeConnected(peerId)`.

The thing is that `PeerStore` doest really have a new or an init proc. This is becaue that type is take from nim-libp2p.

So I would say we have two options:
* a: add a `new` proc for the peerstore and pass all these variables. We would need also our own `PeerStore` extending the nim-libp2p with fields such as maxFailedAttempts. then `canBeConnected` can use these values without having to pass them. but not sure i like having this values both in peer manager and peerstore.
* b: merge peerstore and peermanager so that `canBeConnected` is part of the peermanager, which already has all the fields we need (backofffactors, maxFailedAttempts, etc). or well a simpler approach is to move `canBeConnected` to the peermanager, since now that i think about it, its more related to the peer manager that the peer store.

tldr: what do you thinnk about moving `canBeConnected` to the peermanager, so that all the parameters that are needed dont need to be passed (just peerId)?

File: waku/v2/node/peer_manager/waku_peer_store.nim [hunk 1]
```diff
@@ -55,6 +56,10 @@ proc canBeConnected*(peerStore: PeerStore,
   if failedAttempts == 0:
     return true
 
+  # if there are too many failed attemps, do not reconnect
+  if failedAttempts >= maxFailedAttempts:
+    return false
```
Comment (rymnc) [inline waku/v2/node/peer_manager/waku_peer_store.nim:61]: we should probably add a check earlier to ensure that the maxFailedAttempts is < 14

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 2]
```diff
@@ -235,6 +235,31 @@ proc loadFromStorage(pm: PeerManager) =
...
+  if failedAttempts == 0:
+    return true
+
+  # if there are too many failed attemps, do not reconnect 
+  # the check for 14 comes from overflow described in https://github.com/waku-org/nwaku/issues/1665)
+  if failedAttempts >= pm.maxFailedAttempts or failedAttempts > 14:
+    return false
+
+  # If it errored we wait an exponential backoff from last connection
+  # the more failed attemps, the greater the backoff since last attempt
```
Comment (rymnc) [inline waku/v2/node/peer_manager/peer_manager.nim:255]:
```suggestion
  # the more failed attempts, the greater the backoff since last attempt
```

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 3]
```diff
@@ -235,6 +235,31 @@ proc loadFromStorage(pm: PeerManager) =
...
+  let now = Moment.init(getTime().toUnix, Second)
+  let lastFailed = pm.peerStore[LastFailedConnBook][peerId]
+  let backoff = chronos.seconds(pm.initialBackoffInSec*(pm.backoffFactor^(failedAttempts-1)))
```
Comment (rymnc) [inline waku/v2/node/peer_manager/peer_manager.nim:258]: Hmm.. can we abstract this into a function and fuzz it in the test suite? Will help catch these errors in the futures too

File: tests/v2/test_peer_manager.nim [hunk 4]
```diff
@@ -195,20 +195,14 @@ procSuite "Peer Manager":
...
       # Right after failing there is a backoff period
-      nodes[0].peerManager.peerStore.canBeConnected(
-        nonExistentPeer.peerId,
-        nodes[0].peerManager.initialBackoffInSec,
-        nodes[0].peerManager.backoffFactor) == false
+      nodes[0].peerManager.canBeConnected(nonExistentPeer.peerId) == false
 
     # We wait the first backoff period
-    await sleepAsync(2100.milliseconds)
+    await sleepAsync(2100)
```
Comment (rymnc) [inline tests/v2/test_peer_manager.nim:201]: I think the better practice is to use the unit explicitly, i.e .seconds/.millis etc

File: tests/v2/test_peer_manager.nim [hunk 5]
```diff
@@ -655,3 +649,52 @@ procSuite "Peer Manager":
...
+
+  asyncTest "canBeConnected() returns correct value":
+    let pm = PeerManager.new(
+      switch = SwitchBuilder.new().withRng(rng).withMplex().withNoise()
+      .withPeerStore(10)
+      .withMaxConnections(5)
+      .build(),
+      initialBackoffInSec = 1, # with InitialBackoffInSec = 1 backoffs are: 1, 2, 4, 8secs.
+      backoffFactor = 2,
+      maxFailedAttempts = 10,
```
Comment (rymnc) [inline tests/v2/test_peer_manager.nim:661]: This would test only the happy path, can we randomize these values?
I think tests should be deterministic, but wherever math is we should be fuzzing too

File: tests/v2/test_peer_manager.nim (hunk 4 above)
Comment (vpavlin) [reply tests/v2/test_peer_manager.nim:201]: That makes sense, but when I try to use `milliseconds`, I get this error:

```
/home/vpavlin/devel/github.com/waku-org/nwaku/tests/v2/test_peer_manager.nim(201, 21) Error: type mismatch: got <TimeInterval>
but expected one of:
proc sleepAsync(duration: Duration): Future[void]
  first type mismatch at position: 1
  required type for duration: Duration
  but expression 'milliseconds(2100)' is of type: TimeInterval
proc sleepAsync(ms: int): Future[void]
  first type mismatch at position: 1
  required type for ms: int
  but expression 'milliseconds(2100)' is of type: TimeInterval

expression: sleepAsync(milliseconds(2100))
```

which is why I removed it

File: tests/v2/test_peer_manager.nim (hunk 5 above)
Comment (vpavlin) [reply tests/v2/test_peer_manager.nim:661]: That's a good point, I am testing for failure by modifying the `maxFailedAttempts` on the line 698, but I guess some fuzzing would make sense

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 6]
```diff
@@ -235,6 +235,31 @@ proc loadFromStorage(pm: PeerManager) =
...
+  # wait more before trying again.
+  let failedAttempts = pm.peerStore[NumberFailedConnBook][peerId]
+
+  # if it never errored, we can try to connect
+  if failedAttempts == 0:
+    return true
+
+  # if there are too many failed attemps, do not reconnect 
+  # the check for 14 comes from overflow described in https://github.com/waku-org/nwaku/issues/1665)
+  if failedAttempts >= pm.maxFailedAttempts or failedAttempts > 14:
```
Comment (alrevuelta) [inline waku/v2/node/peer_manager/peer_manager.nim:251]: perhaps i would enforce the chech that `maxFailedAttempts` < some_value when we init the peer manager.

note also that with the current values 14 is ok, but if i set `backoffFactor` to 10000 (nonsense perhaps) i will hit the overflow way earlier.

File: waku/v2/node/peer_manager/peer_manager.nim (hunk 3 above)
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim:258]: not against fuzzing but perhaps overkill? related to my other comment, perhaps we can just assert that the backoff values makes sense when we init the peermanager with initialBackoffInSec and backoffFactor. just run the math in there to ensure values are withing some reasonable limits.

**Review by alrevuelta — CHANGES_REQUESTED** [review]
looking really good, moving it to the peermanager was a great idea. some comments.

File: waku/v2/node/peer_manager/peer_manager.nim (hunk 3 above)
Comment (rymnc) [reply waku/v2/node/peer_manager/peer_manager.nim:258]: that works too, just something we should consider next time :)

File: tests/v2/test_peer_manager.nim (hunk 5 above)
Comment (rymnc) [reply tests/v2/test_peer_manager.nim:661]: This works too, or as @alrevuelta mentioned above, this is a non blocker :)

**Comment by vpavlin** [comment]
@rymnc @alrevuelta I added a check to the `new` proc  - it calculates maximum backoff and checks if it is over 1 week (rather arbitrary value, so we can agree on a different one:) )

This results in either success, exception due to being over 1 week or exception due to overflow. In any case this should prevent the overflow happening later when the node is running.

**Review by alrevuelta — CHANGES_REQUESTED** [review]
looking good, some minor comments and almost ready to go.

can you run a node for some time to ensure it works as expected? if you connect to the existing fleets you should see multiple failed attempts, as many nodes in the network are not reachable (eg behind nat)

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 7]
```diff
@@ -235,6 +242,31 @@ proc loadFromStorage(pm: PeerManager) =
...
+  # It uses an exponential backoff. Each connection attempt makes us
+  # wait more before trying again.
+  let failedAttempts = pm.peerStore[NumberFailedConnBook][peerId]
+
+  # if it never errored, we can try to connect
+  if failedAttempts == 0:
+    return true
+
+  # if there are too many failed attempts, do not reconnect
+  # the check for 14 comes from overflow described in https://github.com/waku-org/nwaku/issues/1665)
```
[the remark this answers is not in the document]
Comment (vpavlin) [reply waku/v2/node/peer_manager/peer_manager.nim:257]: Yeah, I had a check for 14 there initially, that is gone now

File: tests/v2/test_peer_manager.nim (hunk 4 above)
Comment (vpavlin) [reply tests/v2/test_peer_manager.nim]: Ok, I get the error now, glad I looked into Chronos a bit more today:D Fixed

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 8]
```diff
@@ -286,12 +318,20 @@ proc new*(T: type PeerManager,
          maxConnections = maxConnections
     raise newException(Defect, "Max number of connections can't be greater than PeerManager capacity")
 
+  # attempt to calculate max backoff to prevent potential overflows or unreasonably high values
```
[the remark this answers is not in the document]
Comment (alrevuelta) [reply waku/v2/node/peer_manager/peer_manager.nim:320]: just saw the test thanks!

**Review by alrevuelta — APPROVED** [review]
lgtm. thanks for fixing it and addressing the comments!

mind editing the description "as long as maxFailedAttempts is lower than 15 in the default setting"?

**Comment by vpavlin** [comment]
Rewrote the description
